src/Adapter/Profile/Employee/CommandHandler/ResetPasswordHandler.php

```diff
@@ -0,0 +1,165 @@
+     */
+    public function __construct(
+        ConfigurationInterface $configuration,
+        Hashing $hashing,
+        LegacyContext $legacyContext,
+        TranslatorInterface $translator
+    ) {
+        $this->configuration = $configuration;
+        $this->hashing = $hashing;
+        $this->legacyContext = $legacyContext;
+        $this->translator = $translator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(ResetPasswordCommand $command)
+    {
+        if (empty($command->getResetToken()) || empty(trim($command->getEmail()->getValue()))) {
+            throw new ResetPasswordInformationMissingException();
+        }
+
+        $employee = new Employee();
+
+        // Legacy condition. Employee object is loaded in $employee->getByEmail() call, so it must go first.
+        $employeeFound = $employee->getByEmail($command->getEmail()->getValue())
+            && $employee
+            && $employee->id == $command->getEmployeeId()->getValue()
+        ;
+
+        if (!$employeeFound) {
+            throw new EmployeeNotFoundException();
+        }
+
+        $timeExpression = sprintf(
+            '%s + %s minutes',
+            (string) $employee->last_passwd_gen,
+            (string) $this->configuration->get('PS_PASSWD_TIME_BACK')
+        );
+        $canResetPassword = (strtotime($timeExpression) - time()) <= 0;
+
+        if (!$canResetPassword) {
+            throw new PasswordResetTooFrequentException();
+        }
+
+        if ($employee->getValidResetPasswordToken() !== $command->getResetToken()) {
+            throw new ResetPasswordTokenExpiredException();
+        }
+
+        $employee->passwd = $this->hashing->hash($command->getNewPlainPassword());
+        $employee->last_passwd_gen = date('Y-m-d H:i:s', time());
+
+        $params = array(
+            '{email}' => $employee->email,
+            '{lastname}' => $employee->lastname,
+            '{firstname}' => $employee->firstname,
+        );
+
+        $employeeLanguage = new Language((int) $this->legacyContext->getContext()->employee->id_lang);
+
+        $mailWasSent = Mail::send(
+            $employee->id_lang,
+            'password',
+            $this->translator->trans(
+                'Your new password',
+                [],
+                'Emails.Subject',
+                $employeeLanguage->locale
+            ),
+            $params,
+            $employee->email,
+            $employee->firstname . ' ' . $employee->lastname
+        );
+
+        if (!$mailWasSent) {
+            throw new UnableToResetPasswordException();
+        }
+
+        // Update employee only if the mail can be sent
+        Shop::setContext(Shop::CONTEXT_SHOP, (int) min($employee->getAssociatedShops()));
+
+        if (!$employee->update()) {
+            throw new UnableToResetPasswordException();
+        }
+
+        $employee->removeResetPasswordToken(); // Delete temporary reset token
+        $employee->update();
```

Comment:
that's how it was in legacy :thinking: maybe there was some logic behind it

Reply:
@matks @PierreRambaud do we need to call `$employee->update()` twice here?

Legacy logic was here:
https://github.com/PrestaShop/PrestaShop/blob/develop/controllers/admin/AdminLoginController.php#L370

Reply:
I think I understood why it's updating employee twice. I don't know if it's smart though.

First time it's saving the new password and password change date to the database.
Second time (only when the first save succeeded) it removes the token that's used to change the password.

So I guess if the new password could not be saved it would not remove the token from the database and would allow you to initiate password change again. It doesn't look necessary to save twice - we could do it in one go.

wdyt @jolelievre @matks ?

Reply:
Nice finding @rokaszygmantas !

I ... guess the idea is that "we are afraid that something goes wrong in mysql, and the token is removed but the password is not saved".

I prefer to be defensive here: let's keep it as is, but add the informations gathered by Rokas in comments, to explain this.

Reply:
@matks if something goes wrong with the database and the password will not be saved, then the token will not be removed as well - it is the same update query, they are stored in the same table and same record, so they are updated both at the same time